sweepbatcher/greedy_batch_selection.go

```diff
@@ -0,0 +1,308 @@
+// selectBatch if the most cost-efficient action is new batch creation.
+const specialId = -1
+
+// feeDetails is either a batch or a sweep and it holds data important for
+// selection of a batch to add the sweep to (or new batch creation).
+type feeDetails struct {
+	BatchId        int32
+	FeeRate        chainfee.SatPerKWeight
+	CoopWeight     lntypes.WeightUnit
+	NonCoopWeight  lntypes.WeightUnit
+	NonCoopHint    bool
+	IsExternalAddr bool
+}
+
+// fee returns fee of onchain transaction representing this instance.
+func (e feeDetails) fee() btcutil.Amount {
+	var weight lntypes.WeightUnit
+	if e.NonCoopHint {
+		weight = e.NonCoopWeight
+	} else {
+		weight = e.CoopWeight
+	}
+
+	return e.FeeRate.FeeForWeight(weight)
+}
+
+// combine returns new feeDetails, combining properties.
+func (e1 feeDetails) combine(e2 feeDetails) feeDetails {
+	// The fee rate is max of two fee rates.
+	feeRate := e1.FeeRate
+	if feeRate < e2.FeeRate {
+		feeRate = e2.FeeRate
+	}
+
+	return feeDetails{
+		FeeRate:        feeRate,
+		CoopWeight:     e1.CoopWeight + e2.CoopWeight,
+		NonCoopWeight:  e1.NonCoopWeight + e2.NonCoopWeight,
+		NonCoopHint:    e1.NonCoopHint || e2.NonCoopHint,
+		IsExternalAddr: e1.IsExternalAddr || e2.IsExternalAddr,
+	}
+}
+
+// selectBatch returns index of the batch adding to which minimizes costs.
+// For each batch its fee rate and two weight are provided: weight in case of
+// cooperative spending and weight in case non-cooperative spending (using
+// preimages instead of taproot key spend). Also a hint is provided to signal
+// if the batch has to use non-cooperative spending path. The same data is also
+// provided to the sweep for which we are selecting a batch to add. In case of
+// the sweep weights are weight deltas resulted from adding the sweep. Finally,
+// the same data is provided for new batch having this sweep only. The algorithm
+// compares costs of adding the sweep to each existing batch, and costs of new
+// batch creation for this sweep and returns BatchId of the winning batch. If
+// the best option is to create a new batch, it returns specialId. Each fee
+// details has also IsExternalAddr flag. There is a rule that sweeps having flag
+// IsExternalAddr must go in individual batches. Cooperative spending is only
+// available if all the sweeps support cooperative spending path.
+func selectBatch(batches []feeDetails, sweep, oneSweepBatch feeDetails) (int32,
+	error) {
+
+	// Track the best batch to add a sweep to. The default case is new batch
+	// creation with this sweep only in it. The cost is its full fee.
+	bestBatchId := int32(specialId)
+	bestCost := oneSweepBatch.fee()
+
+	// Try to add the sweep to every batch, calculate the costs and
+	// find the batch adding to which results in minimum costs.
+	for _, batch := range batches {
+		// If either the batch or the sweep has IsExternalAddr flag,
+		// the sweep can't be added to the batch, so skip.
+		if batch.IsExternalAddr || sweep.IsExternalAddr {
```

Comment:
In case of `batch.IsExternalAddr` we can't return `specialId` yet. The sweep can be without `IsExternalAddr` and there might be a batch without IsExternalAddr, which it can be added to.

I moved the check against `sweep.IsExternalAddr` out of the cycle, to the beginning of the function.

Renamed `specialId` to `newBatchSignal`.

Reply:
Sounds good!